lib/streamlit/external/langchain/mutable_expander.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Any, NamedTuple, Optional
+
+from streamlit.delta_generator import DeltaGenerator
+from streamlit.type_util import SupportsStr
+
+
+class ChildType(Enum):
+    MARKDOWN = "MARKDOWN"
+    EXCEPTION = "EXCEPTION"
+
+
+class ChildRecord(NamedTuple):
+    type: ChildType
+    kwargs: dict[str, Any]
+    dg: DeltaGenerator
+
+
+class MutableExpander:
```

Comment:
nit: a nice-to-have would be some unit tests for `MutableExpander`. Thought, `StreamlitCallbackHandlerTest ` is covering that on a higher level, so it's not fully necessary. But it would make it easier to potentially reuse `MutableExpander` for other intermediate thought use cases/frameworks as well.